plugins/org.eclipse.elk.alg.vertiflex/src/org/eclipse/elk/alg/vertiflex/p2relative/RelativeXPlacer.java

```diff
@@ -0,0 +1,651 @@
+        myProgressMonitor.done();
+    }
+    
+    /** Computes the distance between two outlines. */
+    private double outlineDistance(final OutlineNode outline1, final OutlineNode outline2) {
+        
+        OutlineNode changedOutline1 = new OutlineNode(outline1.getRelativeX(), MINIMAL_Y, 
+                new OutlineNode(0.0, outline1.getAbsoluteY(), outline1.getNext()));
+        OutlineNode changedOutline2 = new OutlineNode(outline2.getRelativeX(), MINIMAL_Y, 
+                new OutlineNode(0.0, outline2.getAbsoluteY(), outline2.getNext()));
+        
+        // the return value
+        double dist = changedOutline1.getRelativeX() - changedOutline2.getRelativeX();
+        
+        OutlineNode o1;
+        OutlineNode o2;
+        double x1;
+        double x2;
+        double deltaX, deltaY;
+        double newdist;
+        
+        
+        // first run (compare points of o1  with o2)
+        o1 = changedOutline1;
+        o2 = changedOutline2;
+        x1 = o1.getRelativeX();
+        x2 = o2.getRelativeX();
+        while (o1 != null && !o2.isLast()) {
+            if (o2.getNext().getAbsoluteY() > o1.getAbsoluteY()) {
+             // now we compare
+                deltaX = o2.getNext().getRelativeX();
+                deltaY = o2.getNext().getAbsoluteY() - o2.getAbsoluteY();
+                newdist = x1 - x2 - ((o1.getAbsoluteY() - o2.getAbsoluteY()) * deltaX) / deltaY;
+                
+                dist = Math.max(dist, newdist);
+                
+                // now change o1
+                o1 = o1.getNext();
+                if (o1 != null) {
+                    x1 += o1.getRelativeX();
+                }
+            } else {
+                o2 = o2.getNext();
+                x2 += o2.getRelativeX();
+            }
+        }
+        
+        // second run (compare points of o2  with o1)
+        o1 = changedOutline1;
+        o2 = changedOutline2;
+        x1 = o1.getRelativeX();
+        x2 = o2.getRelativeX();
+        while (o2 != null && !o1.isLast()) {
+            if (o1.getNext().getAbsoluteY() > o2.getAbsoluteY()) {
+                // now we compare
+                deltaX = o1.getNext().getRelativeX();
+                deltaY = o1.getNext().getAbsoluteY() - o1.getAbsoluteY();
+                newdist = x1 - x2 + ((o2.getAbsoluteY() - o1.getAbsoluteY()) * deltaX) / deltaY;
+                
+                dist = Math.max(dist, newdist);
+                
+                // now change o2
+                o2 = o2.getNext();
+                if (o2 != null) {
+                    x2 += o2.getRelativeX();
+                }
+            } else {
+                o1 = o1.getNext();
+                x1 += o1.getRelativeX();
+            }
+      
+        }
+        
+        return dist;
+    }
+    
+    /**
+     * This is the recursive function that calculates the layout for one node and it's children.
+     * @param graph 
+     */
+    private void yConsTreeStep(final ElkNode graph) {
```

Comment:
Can this method be split into smaller parts. It looks rather long.
Doing this recursively may be a bad idea.
This method is missing comments what you are trying to do in the separate cases.